File: executor/union_iter.go
// Copyright 2021 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// See the License for the specific language governing permissions and
// limitations under the License.

package executor

import (
	"bytes"

	"github.com/pingcap/tidb/kv"
)

// UnionIter implements kv.Iterator
type UnionIter struct {
	dirtyIt    kv.Iterator
	snapshotIt kv.Iterator

	dirtyValid    bool
	snapshotValid bool

	curIsDirty bool
	isValid    bool
	reverse    bool
}

// NewUnionIter returns a union iterator for BufferStore.
func NewUnionIter(dirtyIt kv.Iterator, snapshotIt kv.Iterator, reverse bool) (*UnionIter, error) {
	it := &UnionIter{
		dirtyIt:       dirtyIt,
		snapshotIt:    snapshotIt,
		dirtyValid:    dirtyIt.Valid(),
		snapshotValid: snapshotIt.Valid(),
		reverse:       reverse,
	}
	err := it.updateCur()
	if err != nil {
		return nil, err
	}
	return it, nil
}

// dirtyNext makes iter.dirtyIt go and update valid status.
func (iter *UnionIter) dirtyNext() error {
	err := iter.dirtyIt.Next()
	iter.dirtyValid = iter.dirtyIt.Valid()
	return err
}

// snapshotNext makes iter.snapshotIt go and update valid status.
func (iter *UnionIter) snapshotNext() error {
	err := iter.snapshotIt.Next()
	iter.snapshotValid = iter.snapshotIt.Valid()
	return err
}

func (iter *UnionIter) updateCur() error {
	iter.isValid = true
	for {
		if !iter.dirtyValid && !iter.snapshotValid {
			iter.isValid = false
			break
		}

		if !iter.dirtyValid {
			iter.curIsDirty = false
			break
		}

		if !iter.snapshotValid {
			iter.curIsDirty = true
			break
		}

		// both valid
		if iter.snapshotValid && iter.dirtyValid {
			snapshotKey := iter.snapshotIt.Key()
			dirtyKey := iter.dirtyIt.Key()
			cmp := bytes.Compare(dirtyKey, snapshotKey)
			if iter.reverse {
				cmp = -cmp
			}
			// if equal, means both have value
			if cmp == 0 {
				if err := iter.snapshotNext(); err != nil {
					return err
				}
				iter.curIsDirty = true
				break
			} else if cmp > 0 {
				// record from snapshot comes first
				iter.curIsDirty = false
				break
			} else {
				// record from dirty comes first
				iter.curIsDirty = true
				break
			}
		}
	}
	return nil
}

// Next implements the Iterator Next interface.
func (iter *UnionIter) Next() error {
	var err error
	if !iter.curIsDirty {
		err = iter.snapshotNext()
	} else {
		err = iter.dirtyNext()
	}
	if err != nil {
		return err
	}
	err = iter.updateCur()
	return err
}

// Value implements the Iterator Value interface.
// Multi columns
func (iter *UnionIter) Value() []byte {
	if !iter.curIsDirty {
		return iter.snapshotIt.Value()
	}
	return iter.dirtyIt.Value()
}

// Key implements the Iterator Key interface.
func (iter *UnionIter) Key() kv.Key {
	if !iter.curIsDirty {
		return iter.snapshotIt.Key()
	}
	return iter.dirtyIt.Key()
}

// Valid implements the Iterator Valid interface.
func (iter *UnionIter) Valid() bool {
	return iter.isValid
}

// Close implements the Iterator Close interface.
func (iter *UnionIter) Close() {
	if iter.snapshotIt != nil {
		iter.snapshotIt.Close()
		iter.snapshotIt = nil
	}
	if iter.dirtyIt != nil {
		iter.dirtyIt.Close()
		iter.dirtyIt = nil
	}
}

File: rust/crates/tidb-txnkv/src/union_iter.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Merged dirty-buffer and snapshot iteration translated from
//! `pkg/store/driver/txn/union_iter.go` and consumed directly by the
//! transaction read driver.

use std::cmp::Ordering;
use std::fmt;

use crate::{Key, KvIterator};

/// Merges a transaction's dirty iterator with its snapshot iterator.
///
/// Both inputs must already be ordered in the direction selected by `reverse`.
/// Dirty values override equal snapshot keys, and an empty dirty value is the
/// source tombstone representation and is never returned to the caller.
pub struct UnionIter<D, S> {
    dirty: Option<D>,
    snapshot: Option<S>,
    dirty_valid: bool,
    snapshot_valid: bool,
    current_is_dirty: bool,
    valid: bool,
    reverse: bool,
}

/// Constructor failure that preserves caller ownership of both inputs.
///
/// Go callers retain their iterator interface values when `NewUnionIter`
/// fails and explicitly close both. Rust moves inputs into constructors, so
/// the error returns them without closing or replacing the originating error.
pub struct UnionIterInitError<D, S, E> {
    error: E,
    dirty: D,
    snapshot: S,
}

impl<D, S, E> UnionIterInitError<D, S, E> {
    /// Borrows the unchanged originating iterator error.
    pub fn error(&self) -> &E {
        &self.error
    }

    /// Returns the error and both still-open input iterators to the caller.
    pub fn into_parts(self) -> (E, D, S) {
        (self.error, self.dirty, self.snapshot)
    }
}

impl<D, S, E: fmt::Debug> fmt::Debug for UnionIterInitError<D, S, E> {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("UnionIterInitError")
            .field("error", &self.error)
            .finish_non_exhaustive()
    }
}

impl<D, S> UnionIter<D, S>
where
    D: KvIterator,
    S: KvIterator<Error = D::Error>,
{
    /// Creates a union iterator at the first visible key.
    ///
    /// As in Go `NewUnionIter`, an error encountered while skipping an initial
    /// tombstone or an equal-key input is returned unchanged. The input
    /// iterators are not closed on this constructor error.
    pub fn new(
        dirty: D,
        snapshot: S,
        reverse: bool,
    ) -> Result<Self, UnionIterInitError<D, S, D::Error>> {
        let dirty_valid = dirty.valid();
        let snapshot_valid = snapshot.valid();
        let mut iterator = Self {
            dirty: Some(dirty),
            snapshot: Some(snapshot),
            dirty_valid,
            snapshot_valid,
            current_is_dirty: false,
            valid: false,
            reverse,
        };
        match iterator.update_current() {
            Ok(()) => Ok(iterator),
            Err(error) => Err(UnionIterInitError {
                error,
                dirty: iterator
                    .dirty
                    .take()
                    .expect("constructor retains the dirty iterator"),
                snapshot: iterator
                    .snapshot
                    .take()
                    .expect("constructor retains the snapshot iterator"),
            }),
        }
    }

    fn dirty(&self) -> &D {
        self.dirty
            .as_ref()
            .expect("dirty iterator is unavailable after close")
    }

    fn dirty_mut(&mut self) -> &mut D {
        self.dirty
            .as_mut()
            .expect("dirty iterator is unavailable after close")
    }

    fn snapshot(&self) -> &S {
        self.snapshot
            .as_ref()
            .expect("snapshot iterator is unavailable after close")
    }

    fn snapshot_mut(&mut self) -> &mut S {
        self.snapshot
            .as_mut()
            .expect("snapshot iterator is unavailable after close")
    }

    fn advance_dirty(&mut self) -> Result<(), D::Error> {
        let result = self.dirty_mut().next();
        self.dirty_valid = self.dirty().valid();
        result
    }

    fn advance_snapshot(&mut self) -> Result<(), D::Error> {
        let result = self.snapshot_mut().next();
        self.snapshot_valid = self.snapshot().valid();
        result
    }

    fn update_current(&mut self) -> Result<(), D::Error> {
        self.valid = true;
        loop {
            if !self.dirty_valid && !self.snapshot_valid {
                self.valid = false;
                break;
            }

            if !self.dirty_valid {
                self.current_is_dirty = false;
                break;
            }

            if !self.snapshot_valid {
                self.current_is_dirty = true;
                if self.dirty().value().is_empty() {
                    self.advance_dirty()?;
                    continue;
                }
                break;
            }

            let mut ordering = self.dirty().key().compare(self.snapshot().key());
            if self.reverse {
                ordering = ordering.reverse();
            }

            match ordering {
                Ordering::Equal => {
                    if self.dirty().value().is_empty() {
                        // Preserve the source error order: dirty advances first,
                        // and snapshot is untouched when that advance fails.
                        self.advance_dirty()?;
                        self.advance_snapshot()?;
                        continue;
                    }
                    // The snapshot advances before the dirty value is exposed,
                    // so the next call advances dirty and cannot repeat the key.
                    self.advance_snapshot()?;
                    self.current_is_dirty = true;
                    break;
                }
                Ordering::Greater => {
                    self.current_is_dirty = false;
                    break;
                }
                Ordering::Less => {
                    if self.dirty().value().is_empty() {
                        self.advance_dirty()?;
                        continue;
                    }
                    self.current_is_dirty = true;
                    break;
                }
            }
        }
        Ok(())
    }

    /// Returns whether the current position contains a visible entry.
    pub fn valid(&self) -> bool {
        self.valid
    }

    /// Returns the current key.
    pub fn key(&self) -> &Key {
        if self.current_is_dirty {
            self.dirty().key()
        } else {
            self.snapshot().key()
        }
    }

    /// Returns the current value.
    pub fn value(&self) -> &[u8] {
        if self.current_is_dirty {
            self.dirty().value()
        } else {
            self.snapshot().value()
        }
    }

    fn advance_union(&mut self) -> Result<(), D::Error> {
        if self.current_is_dirty {
            self.advance_dirty()?;
        } else {
            self.advance_snapshot()?;
        }
        self.update_current()
    }

    /// Closes both inputs. Repeated calls are safe.
    pub fn close(&mut self) {
        if let Some(mut snapshot) = self.snapshot.take() {
            snapshot.close();
        }
        if let Some(mut dirty) = self.dirty.take() {
            dirty.close();
        }
    }
}

impl<D, S> KvIterator for UnionIter<D, S>
where
    D: KvIterator,
    S: KvIterator<Error = D::Error>,
{
    type Error = D::Error;

    fn valid(&self) -> bool {
        UnionIter::valid(self)
    }

    fn key(&self) -> &Key {
        UnionIter::key(self)
    }

    fn value(&self) -> &[u8] {
        UnionIter::value(self)
    }

    fn next(&mut self) -> Result<(), Self::Error> {
        self.advance_union()
    }

    fn close(&mut self) {
        UnionIter::close(self);
    }
}
